pkg/server/file.go

```diff
@@ -114,30 +114,65 @@ func (t *TranslateServiceServer) UploadTranslationFile(
 		return nil, status.Errorf(codes.InvalidArgument, err.Error())
 	}
 
-	switch err := t.repo.SaveMessages(ctx, params.serviceID, messages); {
-	default:
-		// noop
-	case errors.Is(err, repo.ErrNotFound):
-		return nil, status.Errorf(codes.NotFound, "service not found")
-	case err != nil:
-		return nil, status.Errorf(codes.Internal, "")
+	empty, err, done := t.SaveMsgsAndHandleErrors(ctx, params, messages)
+	if done {
+		return empty, err
 	}
 
-	// If the uploaded file is original and populateTranslations flag is true, populate the translated messages.
-	if messages.Original && params.populateTranslations {
-		allMessages, err := t.repo.LoadMessages(ctx, params.serviceID, repo.LoadMessagesOpts{})
+	// When updating original messages, changes might affect translations - transform and update all translations.
+	if messages.Original {
+		all, err := t.repo.LoadMessages(ctx, params.serviceID, repo.LoadMessagesOpts{})
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "")
 		}
 
-		if err = t.populateTranslatedMessages(ctx, params.serviceID, messages, allMessages); err != nil {
-			return nil, err
+		if len(all) == 1 && all[0].Original {
+			return &emptypb.Empty{}, nil
+		}
+
+		prev, _ := all.SplitOriginal()
+
+		// Find original messages with altered text, then replace text in associated messages for all translations.
+		all = t.alterTranslations(all, findUntranslatedIDs(prev, messages))
+
+		// If populateMessages is true - populate missing messages for all translations.
+		if params.populateTranslations {
+			all = t.populateTranslations(all)
+		}
+
+		// Fuzzy translate untranslated messages for all translations
+		if all, err = t.fuzzyTranslate(ctx, all); err != nil {
+			return nil, fmt.Errorf("fuzzy translate messages :%w", err)
+		}
+
+		// Update all translations
+		for i := range all {
+			if all[i].Original {
+				continue
+			}
+
+			empty, err, done = t.SaveMsgsAndHandleErrors(ctx, params, messages)
```

Comment:
There is no need for such a simple function. Any reason you added it? Cognitive complexity?

Reply:
I can see it is reused in a couple of places.